src/platforms/web/runtime/modules/dom-props.js

```diff
@@ -44,6 +47,31 @@ function updateDOMProps (oldVnode: VNodeWithData, vnode: VNodeWithData) {
   }
 }
 
+function isValueChanged (vnode: VNodeWithData, val: string): boolean {
+  const value = vnode.elm.value
+  const modifiers = getModelModifier(vnode)
+  const needNumber = (modifiers && modifiers.number) || vnode.elm.type === 'number'
+  const needTrim = modifiers && modifiers.trim
+  if (needNumber) {
+    return toNumber(value, val) !== toNumber(val)
+  }
+  if (needTrim) {
+    return value.trim() !== val.trim()
+  }
+  return value !== val
+}
+
+function getModelModifier (vnode: VNodeWithData): ASTModifiers | void {
+  const directives = vnode.data.directives || []
+  for (let i = 0, directive; i < directives.length; i++) {
+    directive = directives[i]
+    if (directive.name === 'model') {
+      return directive.modifiers
+    }
+  }
+  return undefined
```

Comment:
Isn't better to return an empty object so at the top you can do `modifiers.number` instead of `modifiers && modifiers.number`?